docs/book/04-reslib.Rmd

```diff
@@ -0,0 +1,187 @@
+# The residue library
+
+Unlike all other PDB loading tools we know about, Pablo exposes its library of known residues to the user for inspection, modification, and even replacement at runtime. This can be done with the [`residue_library`] argument to [`topology_from_pdb`]. `residue_library` takes a mapping from residue IDs to a list of residue definitions (see [](#chapter:resdef)). All of these definitions are matched against any residue with that ID, and all that do match must have consistent chemistry --- Pablo would rather raise an error than provide ambiguous chemistry.
+
+The fact that a residue library admits multiple residue definitions for each name allows alternate protonation states and resonance forms to be included in a residue library. Matches are considered chemically consistent if they have identical elements, connectivity, and net formal charge. Only a few residues have alternative protonation states in default library, so the user will often need to augment it --- thankfully, this is [easy](#protonation-states).
+
+## The default residue library: `STD_CCD_CACHE`
+
+The default value of the [`residue_library`] argument is a module-level constant called [`STD_CCD_CACHE`]. This is an instance of the [`CcdCache`] class that comes pre-loaded with a number of common residues and patches designed to make it more usable for common biomolecular workflows.
+
+The [`CcdCache`] class provides a `Mapping` interface to the CCD itself; unknown residues are downloaded from the CCD when requested, and cached both in memory and on disk in the standard user cache directory. If the `$XDG_CACHE_HOME` environment variable is set, this is `$XDG_CACHE_HOME/openff-pablo/ccd_cache`; otherwise, it defaults to `~/.cache/openff-pablo/ccd_cache` on Linux, and `~/Library/Caches/openff-pablo/ccd_cache` on MacOS.
+
+## Augmenting the default library
+
+The [`CcdCache`] class also provides a fluent, declarative API for easily augmenting the residue library. Unfortunately it's not possible to prepare every residue in the CCD to the standard of, say, the canonical amino acids, and you may encounter bugs and missing protonation states. We try to make it as easy as possible to work around these issues.
+
+If you've generated a fix for a common residue, let us know --- we might upstream it so that everyone can use it!
+
+You can read the entire interface in the [`CcdCache`] reference docs, but we summarize and motivate some important use cases here. In general, the way to add new residue definitions to the default library is with the [`with_`] and [`with_replaced`] methods:
+
+```{python}
+from openff.pablo import STD_CCD_CACHE, ResidueDefinition
+
+my_residue_library = STD_CCD_CACHE.with_([
+    # Create a custom hydrogen peroxide definition
+    ResidueDefinition.from_smiles(
+        "[H:1][O:2][O:3][H:4]",
+        atom_names={1: "H1", 2: "O1", 3: "O2", 4: "H2"},
+        residue_name="HOOH",
+        description="HYDROGEN PEROXIDE",
+    )
+])
```

Comment:
(not blocking) I don't think we've validated our FFs for hydrogen peroxide. As long as it doesn't matter for the example, could we find a different small molecule to use here? (eg. ammonia/methane/etc?)

Reply:
Went with ammonia!